### PR TITLE
[RFC] Introduce vmmops Trait

### PR DESCRIPTION
**Pull request by praveen-pk**
As discussed in #1660 I built vmmops trait for Vm object. While trying to pass the vmmops trait to vcpu in hypervisor module I am running into some challenges.

I captured my questions as comments in this PR.

**Note** this patch set has some compile time errors and doesn't build. I am sending this version out for some feedback.

### CODE REVIEW

**Comment by praveen-pk** [comment]
Seems like I accidentally deleted my comments/questions. Adding them again.

File: vmm/src/vm.rs [hunk 1]
```diff
@@ -442,11 +446,11 @@ impl Vm {
     ) -> Result<Self> {
         #[cfg(target_arch = "x86_64")]
         hypervisor.check_required_extensions().unwrap();
-        let vm = hypervisor.create_vm().unwrap();
+        let vm = hypervisor.create_vm(None).unwrap();
```
Comment (praveen-pk) [inline vmm/src/vm.rs:449]: At the time of creating Vm object in hypervisor, the Vm object in vmm is not fully formed. This forced me to declare vmmops in KvmVM to be Option<Arc<dyn vm::VmmOps>>. This way I set vmmops to **None** first and eventually populate it with correct information.

Does this seem reasonable? Any better way to handle this?

File: vmm/src/vm.rs [hunk 2]
```diff
@@ -295,7 +296,7 @@ impl Vm {
     fn new_from_memory_manager(
         config: Arc<Mutex<VmConfig>>,
         memory_manager: Arc<Mutex<MemoryManager>>,
-        vm: Arc<dyn hypervisor::Vm>,
+        vm: Arc<Mutex<dyn hypervisor::Vm>>,
```
Comment (praveen-pk) [inline vmm/src/vm.rs:299]: Secondly, the logic of setting vmmops to None and eventually setting it to correct Object forced me to change the return type of **create_vm** to be mutable (from hypervisor::Result<Arc> -> Result<Arc<Mutex>>). This change seems sensible.

Now the challenge is changing the inputs for vm in other struct like device_manager, cpu_manager, memory_manager. None of these objects need a Mutable Vm object. But the above change is forcing the vm object passed to these objects to be mutable. Any way I can avoid having to pass mutable Vm object to these Objects?

File: vmm/src/vm.rs (hunk 1 above)
Comment (russell-islam) [reply vmm/src/vm.rs:449]: Does it need to be inlcluded as a member variable in the hypervisor:vm. Can't we just pass the vmmops to the hypervisor functions that need this?

File: hypervisor/src/vm.rs [hunk 3]
```diff
@@ -117,6 +118,26 @@ pub enum HypervisorVmError {
...
+    ///
+    /// Read Guest memory
+    ///
+    #[error("Failed to read guest memory: {0}")]
+    GuestMemRead(#[source] anyhow::Error),
+    ///
+    /// Read from MMIO Bus
+    ///
+    #[error("Failed to read from MMIO Bus: {0}")]
+    MMioBusRead(#[source] anyhow::Error),
```
Comment (rbradford) [inline hypervisor/src/vm.rs:135]: nit: s/MMio/Mmio/

File: hypervisor/src/vm.rs [hunk 4]
```diff
@@ -184,4 +205,15 @@ pub trait Vm: Send + Sync {
...
+    fn set_vmmops(&mut self, vmmops: Arc<dyn VmmOps>) -> vm::Result<()>;
+    /// get vmmops interface
+    fn get_vmmops(&self) -> Result<Arc<dyn VmmOps>>;
+}
+
+pub trait VmmOps: Send + Sync {
+    fn write_guest_mem(&self, buf: &[u8], gpa:u64,) -> vm::Result<usize>;
+    fn read_guest_mem(&self, buf: &mut [u8], gpa:u64,) -> vm::Result<usize>;
+    fn mmio_write(&self, addr: u64, data: &[u8]) -> bool;
+    fn mmio_read(&self, addr: u64, data: &mut [u8]) -> bool;
```
Comment (rbradford) [inline hypervisor/src/vm.rs:218]: I guess you need `io_{read/write}` version for port I/O.

File: vmm/src/vm.rs (hunk 1 above)
Comment (rbradford) [reply vmm/src/vm.rs:449]: Both these seem reasonable to me.

File: vmm/src/vm.rs (hunk 2 above)
Comment (rbradford) [reply vmm/src/vm.rs:299]: I don't think that matters. If there is some contention over the `Mutex` then `RwLock` can peform better (but only when you have multiple regular readers / infrequent writers.

**Review by rbradford — COMMENTED** [review]
I think we need to see this more fleshed out / maybe working to see if this is the right implementation but I think the approach is along the right lines.

**Comment by liuw** [comment]
A general comment for the commit subject and message for each commit: please properly prefix the subject with the component(s) you touch and write in the commit message information related to why you want this change and how you do it. Finally you also need to put in your SoB.

File: hypervisor/src/vm.rs (hunk 4 above)
Comment (praveen-pk) [reply hypervisor/src/vm.rs:218]: yes. I will add that too

**Comment by praveen-pk** [comment]
> A general comment for the commit subject and message for each commit: please properly prefix the subject with the component(s) you touch and write in the commit message information related to why you want this change and how you do it. Finally you also need to put in your SoB.

Will handle these in next version Wei

**Pull request closed without merging**